crates/milli/src/search/new/geo_sort.rs

```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
                 cache.pop_back()
             }
         };
-        while let Some((id, point)) = next(&mut self.cached_sorted_docids) {
-            if geo_candidates.contains(id) {
-                return Ok(Some(RankingRuleOutput {
-                    query,
-                    candidates: RoaringBitmap::from_iter([id]),
-                    score: ScoreDetails::GeoSort(score_details::GeoSort {
-                        target_point: self.point,
-                        ascending: self.ascending,
-                        value: Some(point),
-                    }),
-                }));
+        let put_back = |cache: &mut VecDeque<_>, x: _| {
+            if ascending {
+                cache.push_front(x)
+            } else {
+                cache.push_back(x)
             }
-        }
+        };
 
-        // if we got out of this loop it means we've exhausted our cache.
-        // we need to refill it and run the function again.
-        self.fill_buffer(ctx, &geo_candidates)?;
-        self.next_bucket(ctx, logger, universe)
+        let mut current_bucket = RoaringBitmap::new();
+        // current_distance stores the first point and distance in current bucket
+        // The farthest distance between two points on earth is about 2e7 meters, u32 is big enough to hold any distance.
+        let mut current_distance: Option<([f64; 2], u32)> = None;
+        loop {
+            // The loop will only exit when we have found all points with equal distance or have exhausted the candidates.
+            if let Some((id, point)) = next(&mut self.cached_sorted_docids) {
+                if geo_candidates.contains(id) {
+                    let distance = distance_between_two_points(&self.point, &point).round() as u32;
+                    if let Some((point0, bucket_distance)) = current_distance.as_ref() {
+                        if bucket_distance != &distance {
+                            // different distance, point belongs to next bucket
+                            put_back(&mut self.cached_sorted_docids, (id, point));
+                            return Ok(Some(RankingRuleOutput {
+                                query,
+                                candidates: current_bucket,
+                                score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                    target_point: self.point,
+                                    ascending: self.ascending,
+                                    value: Some(point0.to_owned()),
```

Comment:
Outputting the coordinates of one random points in the bucket is pretty bad and could be misleading.
We should stop returning this value entirely and only output the distance of this bucket from the required point.

But that would be a breaking change and cannot be done in this PR. I’ll open an issue for v2.

Reply:
OK, I'll leave this question alone for now.

By the way, as this [comment](https://github.com/meilisearch/meilisearch/issues/5405#issuecomment-2774762727) says, small differences in distance should be treated as equal. Perhaps we should allow users to customize this difference

Reply:
We should think about it, but it’s more of a product question.

I think your PR is great as meilisearch returns the `_geoDistance` in meter it makes sense that when two documents share the same `_geoDistance` they’re bucketed together 👍